examples/books_dialogroute/lib/main.dart

```diff
@@ -0,0 +1,152 @@
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: Text('Home Screen'),
+      ),
+      body: Center(
+        child: ElevatedButton(
+          onPressed: () => context.beamToNamed('/books'),
+          child: Text('See books'),
+        ),
+      ),
+    );
+  }
+}
+
+class BooksScreen extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: Text('Books'),
+      ),
+      body: ListView(
+        children: books
+            .map(
+              (book) => ListTile(
+                title: Text(book['title']!),
+                subtitle: Text(book['author']!),
+                onTap: () =>
+                    context.beamToNamed('/books/${book['id']}?buy=true'),
+              ),
+            )
+            .toList(),
+      ),
+    );
+  }
+}
+
+class BookDetailsScreen extends StatefulWidget {
+  BookDetailsScreen({
+    required this.bookId,
+  }) : book = books.firstWhere((book) => book['id'] == bookId);
+
+  final String bookId;
+  final Map<String, String> book;
+
+  @override
+  _BookDetailsScreenState createState() => _BookDetailsScreenState();
+}
+
+class _BookDetailsScreenState extends State<BookDetailsScreen> {
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: Text(widget.book['title']!),
+      ),
+      body: Padding(
+        padding: const EdgeInsets.all(8.0),
+        child: GestureDetector(
+          onTap: () => context.currentBeamLocation.update(
+            (state) => state.copyWith(queryParameters: {'buy': 'true'}),
+          ),
+          child: Text('Author: ${widget.book['author']}'),
+        ),
+      ),
+    );
+  }
+}
+
+// LOCATIONS
+class BooksLocation extends BeamLocation {
+  BooksLocation(BeamState state) : super(state);
+
+  @override
+  List<String> get pathBlueprints => ['/books/:bookId'];
+
+  @override
+  List<BeamPage> buildPages(BuildContext context, BeamState state) => [
+        BeamPage(
+          key: ValueKey('home'),
+          child: HomeScreen(),
+        ),
+        if (state.uri.pathSegments.contains('books'))
+          BeamPage(
+            key: ValueKey('books'),
+            child: BooksScreen(),
+          ),
+        if (state.pathParameters.containsKey('bookId'))
+          BeamPage(
+            key: ValueKey('book-${state.pathParameters['bookId']}'),
+            child: BookDetailsScreen(bookId: state.pathParameters['bookId']!),
+          ),
+        if (state.queryParameters.containsKey('buy'))
+          BeamPage(
+              routeBuilder: (RouteSettings settings, Widget child) =>
+                  DialogRoute<void>(
+                      context: context,
+                      builder: (context) => child,
+                      settings: settings),
+              key: ValueKey('book-buy-${state.pathParameters['bookId']}'),
```

Comment:
it's a bit unexpected that `BooketailsScreen` pops when we click outside of dialog to dismiss it. This is due to the `BeamPage.defaultOnPopPage` that will remove the `bookId` path segment.

We need a custom `onPopPage` like this:
```dart
onPopPage: (context, delegate, page) {
  delegate.currentBeamLocation.update(
    (state) => state.copyWith(queryParameters: {}),
  );
  return true;
},
```